icicle/src/vec_ops/vec_ops.cu

```diff
@@ -95,46 +95,95 @@ namespace vec_ops {
+        CHK_IF_RETURN(mont::from_montgomery(d_alloc_vec_a, n, config.ctx.stream, d_alloc_vec_a));
+        is_d_alloc_vec_a_allocated = 1;
+        d_vec_a = d_alloc_vec_a;
+      } else {
+        CHK_IF_RETURN(cudaMallocAsync(&d_alloc_vec_a, n * sizeof(E), config.ctx.stream));
+        CHK_IF_RETURN(cudaMemcpyAsync(d_alloc_vec_a, vec_a, n * sizeof(E), cudaMemcpyHostToDevice, config.ctx.stream));
+        is_d_alloc_vec_a_allocated = 1;
+        d_vec_a = d_alloc_vec_a;
+      }
     } else {
-      d_vec_a = vec_a;
+      if (config.is_input_in_montgomery_form) {
+        CHK_IF_RETURN(cudaMallocAsync(
+          &d_alloc_vec_a, n * sizeof(E), config.ctx.stream)); // Allocate in order not to change the input.
+        CHK_IF_RETURN(mont::from_montgomery(vec_a, n, config.ctx.stream, d_alloc_vec_a));
+        is_d_alloc_vec_a_allocated = 1;
+        d_vec_a = d_alloc_vec_a;
+      } else {
+        d_vec_a = vec_a;
+      }
     }
 
+    int is_d_alloc_vec_b_allocated = 0;
     if (!config.is_b_on_device) {
-      CHK_IF_RETURN(cudaMallocAsync(&d_alloc_vec_b, n * sizeof(E), config.ctx.stream));
-      CHK_IF_RETURN(cudaMemcpyAsync(d_alloc_vec_b, vec_b, n * sizeof(E), cudaMemcpyHostToDevice, config.ctx.stream));
-      d_vec_b = d_alloc_vec_b;
+      if (config.is_input_in_montgomery_form) {
+        CHK_IF_RETURN(cudaMallocAsync(&d_alloc_vec_b, n * sizeof(E), config.ctx.stream));
+        CHK_IF_RETURN(cudaMemcpyAsync(d_alloc_vec_b, vec_b, n * sizeof(E), cudaMemcpyHostToDevice, config.ctx.stream));
+        CHK_IF_RETURN(mont::from_montgomery(d_alloc_vec_b, n, config.ctx.stream, d_alloc_vec_b));
+        is_d_alloc_vec_b_allocated = 1;
+        d_vec_b = d_alloc_vec_b;
+      } else {
+        CHK_IF_RETURN(cudaMallocAsync(&d_alloc_vec_b, n * sizeof(E), config.ctx.stream));
+        CHK_IF_RETURN(cudaMemcpyAsync(d_alloc_vec_b, vec_b, n * sizeof(E), cudaMemcpyHostToDevice, config.ctx.stream));
+        is_d_alloc_vec_b_allocated = 1;
+        d_vec_b = d_alloc_vec_b;
+      }
     } else {
-      d_vec_b = vec_b;
+      if (config.is_input_in_montgomery_form) {
+        CHK_IF_RETURN(cudaMallocAsync(
+          &d_alloc_vec_b, n * sizeof(E), config.ctx.stream)); // Allocate in order not to change the input.
+        CHK_IF_RETURN(mont::from_montgomery(vec_b, n, config.ctx.stream, d_alloc_vec_b));
+        is_d_alloc_vec_b_allocated = 1;
+        d_vec_b = d_alloc_vec_b;
+      } else {
+        d_vec_b = vec_b;
+      }
     }
 
+    int is_d_result_allocated = 0;
     if (!config.is_result_on_device) {
       if (!is_in_place) {
         CHK_IF_RETURN(cudaMallocAsync(&d_result, n * sizeof(E), config.ctx.stream));
+        is_d_result_allocated = 1;
       } else {
         d_result = d_vec_a;
       }
     } else {
       if (!is_in_place) {
         d_result = result;
       } else {
-        d_result = result = d_vec_a;
+        // d_result = result = d_vec_a;    // DEBUG - looks like a bug for in-place.
```

Comment:
@vhnatyk

Reply:
seems not a bug (since there are, I believe, all the tests for all the cases) but a redundant assignment (since it's in-place so `result == vec_a`  and on-device so `d_vec_a = vec_a` earlier). should be `d_result= d_vec_a;` for clarity

Reply:
The test fails with 
d_result= d_vec_a;
and passes with
d_result= result;

Reply:
ohh, unfortunately
```
      if (!is_in_place) {
        d_result = result;
      } else {
        d_result = result;
      }
```
I'm afraid - is logically incorrect 😏  so there must be a bug somewhere else too - still figuring it out, hope to find out asap